ZOOKEEPER-2680:Correct DataNode.getChildren() inconsistent behaviour.

arshadmohammad opened the pull request:
DataNode.getChildren() API behavior should be changed and it should always return empty set if the node does not have any child. Currently this API returns some times null some times empty set.

eribeiro commented:
+1. LGTM. :+1:

afine commented on src/java/test/org/apache/zookeeper/server/DataNodeTest.java:
```diff
@@ -0,0 +1,54 @@
+        assertNotNull(children);
+        assertEquals(0, children.size());
+
+        // Returned empty set must not be modifiable
```
nit: make this a separate test

eribeiro replied:
I usually would agree with separating in two methods, but I think we can leave those in same method as they are so closely related... Aestetically pleasant, I mean. But just my opinion. :smile:

eribeiro replied:
Ignore my previous comment. +1 about putting this in a separate method.

arshadmohammad replied:
Done

eribeiro commented on src/java/test/org/apache/zookeeper/server/DataNodeTest.java:
```diff
@@ -0,0 +1,54 @@
+ */
+package org.apache.zookeeper.server;
+
+import static org.junit.Assert.*;
```
FIY, ZK codebase usually don't put the static import in `Assert` methods, so the methods below are like `Assert.assertNotNull()`

Of course, this is not consistent across the whole codebase and I myself prefer the static import, but usually the wildcard is frowned upon here, afaik.

arshadmohammad replied:
I think it is OK to have one import instead of multiple imports. Not changing any thing here.

eribeiro replied:
Fair enough.

eribeiro commented on src/java/test/org/apache/zookeeper/server/DataNodeTest.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Test
+    public void testGetChildrenRetrunsImmutableEmptySet() {
```
Typo: retruns

eribeiro commented on src/java/test/org/apache/zookeeper/server/DataNodeTest.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Test
+    public void testGetChildrenRetrunsImmutableEmptySet() {
+        // Returned empty set must not be modifiable
```
IMO, the method name makes this comment unnecessary

eribeiro commented:
@arshadmohammad My two last questions (really!):
 
* Is this change supposed to be applied to branch-3.4 and branch-3.5? 

* I see that `DataNode.getChildren()` is called in half a dozen places with more or less the logic below:

```
            Set<String> childs = node.getChildren();
            if (childs != null) {
                children = childs.toArray(new String[childs.size()]);
            }
            // children size is zero

           if (children != null) {
               for (String child : children) {
                   // DO LOGIC
               }
           }
```

As `children` is zero length the for-each loop is not executed. I am fine with leaving this code as-is. Otherwise, we could remove the now useless `if (children != null)` calls. Wdyt?

arshadmohammad commented:
1. yes, changes should be applied to branch-3.4 and branch-3.5 also. I will raise merge request for branch-3.4 and branch-3.5 after it is committed to master
2. This is very much needed. Thanks :-). I removed Null check from all references of getChildren and corrected the code as per the need

arshadmohammad commented:
checking the CI failure.

edwardoliveira commented on src/java/main/org/apache/zookeeper/server/DataTree.java:
```diff
@@ -1045,18 +1046,16 @@ private void getCounts(String path, Counts counts) {
         synchronized (node) {
-            children = node.getChildren();
+            Set<String> childs = node.getChildren();
+            children = childs.toArray(new String[childs.size()]);
```
FYI (only): I *guess* some tests broke 'cause `toArray` copies the contents of `childs` to `children` while `node.getChildren()` returns a internal reference (subject to underneath change) to `children`. So, it would need to be something like `children = new HashSet<>(childs);`

eribeiro commented on src/java/main/org/apache/zookeeper/server/DataTree.java:
```diff
@@ -1045,18 +1046,16 @@ private void getCounts(String path, Counts counts) {
         int len = 0;
         synchronized (node) {
-            children = node.getChildren();
+            Set<String> childs = node.getChildren();
```
@arshadmohammad FYI (only): I guess some tests broke 'cause toArray copies the contents of childs to children while node.getChildren() returns a internal reference (subject to underneath change) to children. So, it would need to be something like children = new HashSet<>(childs);

eribeiro replied:
But as all tests are passing, that's fine. :smiley: Congrats!

hanm commented:
+1, merging to master (the jira mentioned patch to 3.4 and 3.5 will be sent separately).
